File: day13/day13.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <tuple>
#include <unordered_set>
#include <vector>

// Using abseil for error-handling and string utilities. Compile with
// 	g++ -std=c++20 -o day13 day13.cpp -l absl_status -l absl_statusor -l absl_strings
#include <absl/status/statusor.h>
#include <absl/strings/numbers.h>
#include <absl/strings/str_split.h>
#include <absl/strings/string_view.h>

enum Axis {
  AxisX,
  AxisY,
};

struct Fold {
  Axis on;
  int value;
};

typedef std::pair<int, int> Point;

struct PointHash {
  std::size_t operator()(const Point &p) const {
    return std::hash<int>()(p.first) ^ std::hash<int>()(p.second);
  }
};

typedef std::unordered_set<Point, PointHash> PointSet;

struct Data {
  PointSet points;
  std::vector<Fold> folds;
};

absl::StatusOr<Data> read(const std::string &filename) {
  Data data;

  std::ifstream file;
  try {
    file.open(filename);
    std::string line;
    bool found_sep = false;
    while (std::getline(file, line)) {
      if (line.empty()) {
        found_sep = true;
        break;
      }
      std::vector<absl::string_view> v = absl::StrSplit(line, ",");
      if (v.size() != 2) {
        return absl::InvalidArgumentError("file contains invalid point");
      }
      int x, y;
      if (!absl::SimpleAtoi(v[0], &x)) {
        return absl::InvalidArgumentError("file contains invalid point");
      }
      if (!absl::SimpleAtoi(v[1], &y)) {
        return absl::InvalidArgumentError("file contains invalid point");
      }
      data.points.insert(std::make_pair(x, y));
    }
    if (!found_sep) {
      return absl::InvalidArgumentError("file contains no empty line");
    }
    while (std::getline(file, line)) {
      auto l = absl::string_view(line);
      if (!absl::ConsumePrefix(&l, "fold along ")) {
        return absl::InvalidArgumentError("fold instruction without prefix");
      }
      Fold f;
      std::vector<absl::string_view> v = absl::StrSplit(l, "=");
      if (v.size() != 2) {
        return absl::InvalidArgumentError(
            "fold instruction without assignment");
      }
      if (v[0] == "x") {
        f.on = AxisX;
      } else if (v[0] == "y") {
        f.on = AxisY;
      } else {
        return absl::InvalidArgumentError("fold instruction with invalid axis");
      }
      if (!absl::SimpleAtoi(v[1], &f.value)) {
        return absl::InvalidArgumentError(
            "fold instruction with invalid value");
      }
      data.folds.push_back(f);
    }
  } catch (...) {
    return absl::InternalError("could not read file");
  };
  return data;
}

Point apply_to_point(Point p, Fold f) {
  switch (f.on) {
  case AxisX:
    return std::make_pair(p.first < f.value ? p.first : 2 * f.value - p.first,
                          p.second);
  case AxisY:
    return std::make_pair(p.first, p.second < f.value ? p.second
                                                      : 2 * f.value - p.second);
  }
  throw std::runtime_error("invalid folding axis");
}

void apply(PointSet *points, Fold f) {
  std::vector<Point> next;
  for (auto p : *points) {
    next.push_back(apply_to_point(p, f));
  }
  points->clear();
  for (auto p : next) {
    points->insert(p);
  }
}

void print_paper(const PointSet &points) {
  int mx = 0;
  int my = 0;
  for (auto p : points) {
    mx = p.first > mx ? p.first : mx;
    my = p.second > my ? p.second : my;
  }
  Point p;
  for (p.second = 0; p.second <= my; p.second++) {
    for (p.first = 0; p.first <= mx; p.first++) {
      if (points.contains(p)) {
        std::cout << "•";
      } else {
        std::cout << " ";
      }
    }
    std::cout << std::endl;
  }
}

int main() {
  auto result = read("input.txt");
  if (!result.ok()) {
    std::cerr << result.status() << std::endl;
    return 1;
  }
  auto data = *result;

  apply(&data.points, data.folds.at(0));
  std::cout << "After first fold, there are " << data.points.size() << " points"
            << std::endl;
  for (int i = 1; i < data.folds.size(); i++) {
    apply(&data.points, data.folds.at(i));
  }
  std::cout << "After all folds, the paper looks like:" << std::endl;
  print_paper(data.points);
}

File: day13/day13.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"log"
	"os"
	"strconv"
	"strings"
)

func main() {
	points, folds, err := parse(os.Stdin)
	if err != nil {
		log.Fatal(err)
	}
	points.Apply(folds[0])
	fmt.Printf("After first fold, there are %d points\n", len(points))
	for _, f := range folds[1:] {
		points.Apply(f)
	}
	fmt.Println("After all folds the paper looks like:")
	points.PrettyPrint()
}

type Axis int

const (
	AxisX Axis = iota
	AxisY
)

type Fold struct {
	Axis  Axis
	Value int
}

type Point struct {
	X int
	Y int
}

func (p Point) Apply(f Fold) Point {
	switch f.Axis {
	case AxisX:
		if p.X > f.Value {
			return Point{2*f.Value - p.X, p.Y}
		}
	case AxisY:
		if p.Y > f.Value {
			return Point{p.X, 2*f.Value - p.Y}
		}
	}
	return p
}

func (p Point) String() string {
	return fmt.Sprintf("(%d,%d)", p.X, p.Y)
}

type PointSet map[Point]struct{}

func (s PointSet) Contains(p Point) bool {
	_, ok := s[p]
	return ok
}

func (s PointSet) Add(p Point) {
	s[p] = struct{}{}
}

func (s PointSet) Remove(p Point) {
	delete(s, p)
}

func (s PointSet) Apply(f Fold) {
	var additions []Point
	for p := range s {
		q := p.Apply(f)
		if p != q {
			delete(s, p)
			additions = append(additions, q)
		}
	}
	for _, p := range additions {
		s.Add(p)
	}
}

func (s PointSet) PrettyPrint() {
	mx, my := -1, -1
	for p := range s {
		if p.X > mx {
			mx = p.X
		}
		if p.Y > my {
			my = p.Y
		}
	}
	for y := 0; y <= my; y++ {
		for x := 0; x <= mx; x++ {
			if s.Contains(Point{x, y}) {
				fmt.Print("•")
			} else {
				fmt.Print(" ")
			}
		}
		fmt.Println()
	}
}

func parse(r io.Reader) (PointSet, []Fold, error) {
	points := make(PointSet)
	var folds []Fold

	s := bufio.NewScanner(r)
	for s.Scan() {
		if s.Text() == "" {
			break
		}
		sp := strings.Split(s.Text(), ",")
		if len(sp) != 2 {
			return nil, nil, errors.New("invalid point")
		}
		x, err := strconv.Atoi(sp[0])
		if err != nil {
			return nil, nil, err
		}
		y, err := strconv.Atoi(sp[1])
		if err != nil {
			return nil, nil, err
		}
		p := Point{x, y}
		if points.Contains(p) {
			return nil, nil, fmt.Errorf("duplicate point %v", p)
		}
		points.Add(p)
	}
	for s.Scan() {
		l := s.Text()
		if !strings.HasPrefix(l, "fold along ") {
			return nil, nil, errors.New("invalid fold instruction")
		}
		l = strings.TrimPrefix(l, "fold along ")
		sp := strings.Split(l, "=")
		val, err := strconv.Atoi(sp[1])
		if err != nil {
			return nil, nil, err
		}
		switch sp[0] {
		case "x":
			folds = append(folds, Fold{AxisX, val})
		case "y":
			folds = append(folds, Fold{AxisY, val})
		default:
			return nil, nil, fmt.Errorf("invalid axis %q", sp[0])
		}
	}
	return points, folds, s.Err()
}
